Add scantxoutset

sgeisler opened the pull request:
Allows to query the bitcoin node's UTXO set for UTXOs belonging to descriptors.

shesek commented on json/src/lib.rs:
```diff
@@ -1140,6 +1140,39 @@ pub enum PubKeyOrAddress<'a> {
+
+#[derive(Serialize, Deserialize, Clone, PartialEq, Eq, Debug)]
+pub struct ScanUtxoResult {
+    pub success: bool,
```
The `success`, `txouts`, `height` and `bestblock` fields didn't seem to exist [on 0.18](https://bitcoincore.org/en/doc/0.18.0/rpc/blockchain/scantxoutset/), which is supported by bitcoincore-rpc iiuc.

sgeisler replied:
fixed

shesek commented on json/src/lib.rs:
```diff
@@ -1140,6 +1140,39 @@ pub enum PubKeyOrAddress<'a> {
+    pub success: bool,
+    pub txouts: u64,
+    pub height: u64,
+    pub bestblock: bitcoin::BlockHash,
```
I'm not quite sure about the naming convention for `bestblock`. Names are usually converted to snake_case, but `GetTxOutResult` uses `bestblock`, while `GetBlockchainInfoResult` uses `best_block_hash`.

sgeisler replied:
fixed

shesek commented on json/src/lib.rs:
```diff
@@ -1140,6 +1140,39 @@ pub enum PubKeyOrAddress<'a> {
+#[derive(Serialize, Deserialize, Clone, PartialEq, Eq, Debug)]
+pub struct ScanUtxoResult {
+    pub success: bool,
+    pub txouts: u64,
```
In #132 I also had a `txouts` field which I rename to `tx_outs`. I chose this because of how `get_tx_out()` and `GetTxOutResult` are called, with `tx` and `out` as separate words. If we decide that `txouts` is preferred, #132 should be changed accordingly to be consistent.

shesek replied:
Ah, I also had a `bestblock` in #132, which I renamed to `best_block`. Same goes for that.

shesek commented on client/src/client.rs:
```diff
@@ -957,6 +959,10 @@ pub trait RpcApi: Sized {
     }
+
+    fn scan_txout_set(&self, descriptors: &[ScanTxoutRequest]) -> Result<json::ScanUtxoResult> {
+        self.call("scantxoutset", &["start".into(), into_json(descriptors)?])
```
What about the `abort` and `status` actions?

If this is only meant to support starting a scan, maybe rename this to something like `scan_txout_set_start()`?

sgeisler replied:
I'm super confused by these actions. We model `start` as a blocking action, so why enable abort? If you see a use case I'll add two functions for these, but I didn't see the use case.

shesek replied:
I'm not sure really, but if its supported in Core then there probably is some merit to it and I think that bitcoincore-rpc should ideally support it too.

But at the very least, if the RpcApi method provides a subset of the functionality of the RPC command, it should not use the same name. I think its alright to add support just for `start` for now, if we just give the method a name that hinted at that.

sgeisler replied:
I guess I could just add `_blocking` to the name to indicate it doesn't support the (probably?) somewhat async API.

sgeisler replied:
fixed by renaming

shesek replied:
The `_blocking` is a bit odd because all methods are blocking, this might create the impression that non-`_blocking` methods are not blocking.

shesek commented on json/src/lib.rs:
```diff
@@ -1140,6 +1140,39 @@ pub enum PubKeyOrAddress<'a> {
+    Single(String),
+    Extended {
+        dest: String,
+        range: u64,
```
`range` can also be a tuple of `(start_index,end_index)`.

sgeisler replied:
I guess I'll have to impl a custom serializer for that as it has either to be a number _or_ a array with two numbers (super annoying). Alternatively we could require a lower limit.

shesek replied:
I think that requiring a lower limit is reasonable, adding an explicit `0` start range is not much of an hassle and allows to keep things simpler.

sgeisler replied:
Fixed, made it a tuple. Maybe it should actually be `(u23, u32)` though? Afaik key derivation can only go up to `2^31-1` for each hardened and unhardened keys according to bip32. So u64 is a bit too big, but u32 too and I don't really want to get into using `ChildNumber` here.

stevenroose commented:
Can you rebase? And perhaps rename that variable to descriptor? :D Unless you think "desc" is a commonly accepted abbreviation for descriptor in the RPC API.

sgeisler commented:
@stevenroose I rebased and fixed the integration test by moving it up in the execution list (idk what the problem was exactly, something about the wallet not being selected).

shesek commented:
What about the `Txout` vs `TxOut` and `txout` vs `tx_out` naming?

We currently have `GetTxOutResult` and `GetTxOutSetInfoResult`, and no occurrences of `Txout`.

For lower case, we have `get_tx_out()`, `get_tx_out_proof()`, `get_tx_out_set_info()` and `GetTxOutSetInfoResult::tx_outs`, and no occurrences of `txout`.

sgeisler commented:
Sorry, I overlooked that. While I don't like the emphasis on the *out* I agree it should follow project conventions.
